### PR TITLE
RFC: WIP: Make macro hygiene easier to use and less error-prone

### PR DESCRIPTION
**Pull request by vtjnash**
This is a rebase and squash (plus updates) of PR #6910

TODO items:
- [ ] handle `new` in type definitions in macros
- [ ] handle fieldnames in type definitions in macros
- [ ] move export of `@hygienic` to exports.jl
- [ ] re-import doc changes from original PR
- [ ] solicit feedback

### CODE REVIEW

File: test/macros.jl [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+@test Which_test.test1(123) == Tuple{Int}
+
+# Should get error("no method found for the specified argument types")
+# in which(), not UnexpectedException
+@test_throws ErrorException Which_test.test1("foo")
+
+# Test gen_call_with_extracted_types's alternate path for keyword arguments
+@test Which_test.test2(1.0) == Tuple{Real}
+@test Which_test.test2(123) == Tuple{Int}
+
```
Comment (tkelman) [inline test/macros.jl]: whitespace here

also would've been nice to preserve `@david-moon`s git authorship here

**Comment by JeffBezanson** [comment]
I think this needs to be merged early in a release cycle.

File: test/macros.jl (hunk 1 above)
Comment (mauro3) [inline test/macros.jl]: To preserve authorship this approach can be used: http://blog.spreedly.com/2014/06/24/merge-pull-request-considered-harmful/

Comment (vtjnash) [inline test/macros.jl]: i can also run `git commit --amend --author <original author>` to fix this up, and had intended to do so before pushing. thanks for the reminder.

**Comment by tkelman** [comment]
Sounds familiar https://github.com/JuliaLang/julia/pull/6910#issuecomment-44046105

Packages that make heavy use of macros (JuMP as an especially notable example) might need to make major changes to adapt to this. This would be a good candidate to run PkgEvaluator on prior to merging.

**Comment by pao** [comment]
Should the milestone get pushed out to 0.5, so there's some chance this doesn't slip through the cracks in (fingers crossed) a few months?

**Comment by mlubin** [comment]
Well, if we merge it soon then it will break JuMP, we'll fix it, and then we'll be able to quickly drop the old-style code after the 0.4 release. If we merge it at the start of the 0.5 cycle, it'll be less disruptive for users but we'll have to keep maintaining the old-style code for about a year until the 0.5 release. I'm ambivalent as to which is better, but either way please keep us in the loop and give us a chance to evaluate this before merging.

**Comment by JeffBezanson** [comment]
I'm not convinced that this is a large improvement. `@hygienic` introduces a strange lack of referential transparency, where it suddenly matters *where* your quote expressions are, and not just what expressions they yield. You also end up wrapping some expressions in `:hygienic`. That may be better in some ways, but at the end of the day it's very similar to how we have to wrap some things in `:escape` now. My impression is that this is roughly as fiddly as what we have now.

**Comment by vtjnash** [comment]
I think the argument is that fiddling with `esc` is non-intuitive. Whereas this actually interprets expressions in macros using rules that are closer to the context they are used in.

My understanding from the original PR is that `@hygienic` is a bit of a side-effect from not adding support to eval for unwrap hygienic Exprs, but that it too can go away with a bit more work.

**Comment by vtjnash** [comment]
@mlubin do you have any sense for how badly this would break JuMP? one of the original goals of the PR was to be as non-disruptive to existing codebases as possible.

**Comment by JeffBezanson** [comment]
> a side-effect from not adding support to eval for unwrap hygienic Exprs, but that it too can go away with a bit more work

Then I think we should really not rush this, and take the time to do it right. My gut feel is that we are at capacity with things to worry about right now.

**Comment by vtjnash** [comment]
sounds good. i wanted to rebase this, in part, just so it wouldn't get too far behind

**Comment by JeffBezanson** [comment]
Yes, it's very nice to have it rebased.

**Comment by toivoh** [comment]
Agreed, let's not rush this but think through how we really want things to work.

Having written macros whose implementations effectively span multiple modules I must say though, that the hygiene aspect that depends on where the quote block is in the code is highly compelling; I've basically had to fall back to the old pre-hygiene tricks in a number of such cases.

**Comment by JeffBezanson** [comment]
@david-moon @toivoh I wonder if hygienic `quote` should resolve globals to the module containing the quote? I suppose to implement that we'd have to mark symbols as `(hygienic ,module ,sym)`?

**Comment by david-moon** [comment]
@david-moon @toivoh I wonder if hygienic quote should resolve globals to the module containing the quote? I suppose to implement that we'd have to mark symbols as (hygienic ,module ,sym) ? 
— 

Yes, symbols in the mark expansion that come from the macro definition rather than from the macro call should be looked up in the environment (including module) where the macro was defined. Did I forget to do that part when I prototyped this?

**Comment by JeffBezanson** [comment]
We seem to be missing the manual section that was written for this.

**Comment by tkelman** [comment]
> Should the milestone get pushed out to 0.5, so there's some chance this doesn't slip through the cracks in (fingers crossed) a few months?

1 year later, this happened. Don't think this is still 0.5.

**Comment by stevengj** [comment]
Any update on this? Would be good to have hygiene finalized for 1.0.

**Comment by StefanKarpinski** [comment]
I think that a system like Clojure's with fully scoped symbols could be the simplest. It wouldn't even need to be particularly breaking since symbol literals like `:foo` could be "relative symbols" while if you wrote `:(a + b)` in a module `M` you would implicitly have `M.a` and `M.b` symbols. Of course, this forces us to stop punning on `:a` versus `:(a)` but I think that's for the best anyway.

**Comment by JeffBezanson** [comment]
I'm not sure about that. It pretty drastically changes what a symbol is. How does it work out for, say, keyword arguments and field names?

**Comment by StefanKarpinski** [comment]
That's a good question. I guess we should look at what Clojure does.

**Comment by StefanKarpinski** [comment]
If you continue to consider symbols to be implicitly relative but introduce a notion of fully qualified identifiers then keyword arguments and field names could continue to be represented by symbols (implicitly relative) while when you write something like `:(a)` it would be resolved to `:(M.a)`. Maybe this isn't much different than what this PR does anyway.

**Comment by StefanKarpinski** [comment]
@david-moon sent me a message:

---

Stefan Karpinski wrote:

> If you continue to consider symbols to be implicitly relative but introduce a notion of fully qualified identifiers then keyword arguments and field names could continue to be represented by symbols (implicitly relative) while when you write something like :(a) it would be resolved to :(M.a). Maybe this isn't much different than what this PR does anyway.

Yes, I think your choices for properly hygienic macros are either to draw a distinction between symbols (run-time objects) and identifiers (compile-time objects), or to do what I did (assuming I remember correctly, it's been years) and allow symbols to be interned in a dictionary specific to a particular macro expansion as well as in the usual global dictionary. This prevents symbols introduced by a macro from being eq to symbols in the macro call, but it requires normalizing those local symbols to global symbols when they end up as literals, keyword arguments, field names, etc. in the macro expansion. When those local symbols end up as bound variables, everything just works. When they end up as unbound variables, they need to be looked up in the scope where the macro was defined rather than in the scope where it was called.

Whatever you do, it shouldn't have a built-in restriction that a macro can only be defined in module scope, not in a scope nested inside a module.  Even if for implementation reasons you need that restriction at first, it shouldn't be built deeply into the language.  An identifier introduced by a macro should be in the scope where the macro was defined regardless of the nature or lifetime of that scope.

[I have a feeling this email reply is going to be prevented from going to the correct mailing list.  Could you please forward it?]

--Dave Moon

**Comment by JeffBezanson** [comment]
For globals, we do have a representation for fully-qualified identifiers: `GlobalRef` objects. I'm not sure it will help much to resolve `:(a)` to `:(M.a)` (a `GlobalRef`) very early, since you still need the surrounding context to know what the symbol actually refers to. But it does seem likely some version of this is similar to what's here, especially if we include https://github.com/JuliaLang/julia/pull/10940#issuecomment-166011360

**Comment by stevengj** [comment]
I guess this will not make 0.6, but it would be nice to have for 1.0

**Comment by vtjnash** [comment]
don't think this helpful anymore to have open, future work can just as easily start from the original (#6910)

**Comment by mauro3** [comment]
Should the original be re-opened?  And its milestone updated to 2.0?

**Pull request closed without merging**